src/ioctls/vm.rs

```diff
@@ -380,6 +380,125 @@ impl VmFd {
+    /// # use kvm_bindings::{kvm_userspace_memory_region, KVM_MEM_LOG_DIRTY_PAGES};
+    /// # let kvm = Kvm::new().unwrap();
+    /// # let vm = kvm.create_vm().unwrap();
+    /// let mem_size = 0x10000;
+    /// let guest_addr: u64 = 0x10000;
+    /// let load_addr: *mut u8 = unsafe {
+    ///     libc::mmap(
+    ///         null_mut(),
+    ///         mem_size,
+    ///         libc::PROT_READ | libc::PROT_WRITE,
+    ///         libc::MAP_ANONYMOUS | libc::MAP_SHARED | libc::MAP_NORESERVE,
+    ///         -1,
+    ///         0,
+    ///     ) as *mut u8
+    /// };
+    ///
+    /// // Initialize a guest memory region using the flag `KVM_MEM_LOG_DIRTY_PAGES`.
+    /// let mem_region = kvm_userspace_memory_region {
+    ///     slot: 0,
+    ///     guest_phys_addr: guest_addr,
+    ///     memory_size: mem_size as u64,
+    ///     userspace_addr: load_addr as u64,
+    ///     flags: KVM_MEM_LOG_DIRTY_PAGES,
+    /// };
+    /// vm.set_user_memory_region(mem_region).unwrap();
+    ///
+    /// let aarch64_code = [
+    ///     0x01, 0x00, 0x00, 0x10, /* adr x1, <this address> */
+    ///     0x22, 0x10, 0x00, 0xb9, /* str w2, [x1, #16]; write to this page */
+    ///     0x02, 0x00, 0x00, 0xb9, /* str w2, [x0]; force MMIO exit */
+    ///     0x00, 0x00, 0x00, 0x14, /* b <this address>; shouldn't get here, but if so loop forever */
+    /// ];
+    ///
+    /// // Write the code in the guest memory. This will generate a dirty page.
+    /// unsafe {
+    ///     let mut slice = slice::from_raw_parts_mut(load_addr, mem_size);
+    ///     slice.write(&aarch64_code).unwrap();
+    /// }
+    ///
+    /// let vcpu_fd = vm.create_vcpu(0).unwrap();
+    /// let mut kvi = kvm_bindings::kvm_vcpu_init::default();
+    /// vm.get_preferred_target(&mut kvi).unwrap();
+    /// vcpu_fd.vcpu_init(&kvi).unwrap();
+    ///
+    /// let core_reg_base: u64 = 0x6030_0000_0010_0000;
+    /// let mmio_addr: u64 = guest_addr + mem_size as u64;
+    /// vcpu_fd.set_one_reg(core_reg_base + 2 * 32, guest_addr); // set PC
+    /// vcpu_fd.set_one_reg(core_reg_base + 2 * 0, mmio_addr); // set X0
+    ///
+    /// loop {
+    ///     match vcpu_fd.run().expect("run failed") {
+    ///         VcpuExit::MmioWrite(addr, data) => {
+    ///             // The code snippet dirties its own page.
+    ///             let dirty_pages_bitmap = vm.get_dirty_log(0, mem_size).unwrap();
+    ///             let dirty_pages = dirty_pages_bitmap
+    ///                     .into_iter()
+    ///                     .map(|page| page.count_ones())
+    ///                     .fold(0, |dirty_page_count, i| dirty_page_count + i);
+    ///             assert_eq!(dirty_pages, 1);
+    ///             break;
+    ///         }
+    ///         exit_reason => panic!("unexpected exit reason: {:?}", exit_reason),
+    ///     }
+    /// }
+    /// ```
+    ///
+    #[cfg(target_arch = "aarch64")]
+    pub fn get_dirty_log(&self, slot: u32, memory_size: usize) -> Result<Vec<u64>> {
```

Comment:
This function looks the same as the x86_64 variant. If it is suppose to work the same on both x86 and aarch64, we can just add the aarch64 to the target_arch macro.

I was living with the impression that the `KVM_GET_DIRTY_LOG` ioctl is only available on x86. Check this issue for details: https://github.com/firecracker-microvm/firecracker/issues/866. I am very surprised to see it work on aarch64 as well.

Reply:
> This function looks the same as the x86_64 variant. If it is suppose to work the same on both x86 and aarch64, we can just add the aarch64 to the target_arch macro.

I would be happy to share the code. The problem is, that without introducing wrapper functions, I didn't see a way to share it since the unit tests above them in the '///' are arch specific.

I would like to learn how to share vmm code among different architectures that need different unit tests, and arch-neutral unit test boilerplate code among those different unit tests.

> 
> I was living with the impression that the `KVM_GET_DIRTY_LOG` ioctl is only available on x86. Check this issue for details: [firecracker-microvm/firecracker#866](https://github.com/firecracker-microvm/firecracker/issues/866). I am very surprised to see it work on aarch64 as well.

KVM_GET_DIRTY_LOG is supported by all architectures that support KVM. I see the documentation (Documentation/virtual/kvm/api.txt) only states x86, but that's a doc bug.

Reply:
So it seems I am not very good at explaining. I modified your code a bit so that we do not have two functions with same code.
You can annotate parts of the code as architecture specific using macros. I did this in a commit on top of your existing ones. Check it out here: https://github.com/andreeaflorescu/kvm-ioctls/blob/test_run_code_archc/src/ioctls/vm.rs#L260

Basically, the only things that are platform specific right now are:
- setting up the registers
- initialize the vcpu (only on aarch64)
- and the VM exit where we check the dirty log. 

Btw, isn't it possible to have a dummy code similar like for x86_64 that only triggers the halt VM exit? I am asking because the example code could be further simplified if we do that as we wouldn't need to trigger two different VM exits to check dirty log.

Reply:
This is the diff between your last commit and mine to help visualize things: https://github.com/rust-vmm/kvm-ioctls/commit/8d0d62b7d7955157567c9676ef2e854a69e1fe24

Reply:
> Check it out here: https://github.com/andreeaflorescu/kvm-ioctls/blob/test_run_code_archc/src/ioctls/vm.rs#L260

That looks pretty good, but I'm not sure interleaving x86 and arm code is the most readable. Now that I know how to use the annotations I might try some refactoring and then create exactly one arm sequence and one x86 sequence.

> Btw, isn't it possible to have a dummy code similar like for x86_64 that only triggers the halt VM exit? I am asking because the example code could be further simplified if we do that as we wouldn't need to trigger two different VM exits to check dirty log.

arm doesn't have a halt exit reason, nor anything similar. Since x86 does also exit on mmio we could change the x86 code to use the same exit as arm though.

Reply:
We can also have x86_64 and arm examples separate. It just felt to me that a lot of code is similar so I didn't feel the need to split it further more. It's your choice of course.

We can use MmioWrite for both platforms, I just found the MmmioWrite to be a bit unrelated to what we actually want to test. I was looking for either halt or nop.